mapie/calibrators/base.py

```diff
@@ -0,0 +1,85 @@
+from __future__ import annotations
+
+from abc import ABCMeta, abstractmethod
+from typing import List, Optional
+
+from sklearn.base import BaseEstimator
+
+from mapie._typing import ArrayLike, NDArray
+
+
+class BaseCalibrator(BaseEstimator, metaclass=ABCMeta):
+    """
+    Base abstract class for the calibrators.
+
+    The ``BaseCalibrator`` subclasses should have at least two methods:
+
+    - ``fit`` : Fit the calibrator to estimate the conformity scores
+        quantiles.
+
+    -  ``predict`` : Predict the conformity score quantiles.
+
+    Attributes
+    ----------
+    fit_attributes: Optional[List[str]]
+        Name of attributes set during the ``fit`` method, and required to call
+        ``predict``.
+    """
+
+    fit_attributes: List[str]
+    sym: bool
+    alpha: Optional[float]
+    random_state: Optional[int]
+
+    @abstractmethod
+    def fit(
+        self,
+        X_calib: ArrayLike,
+        conformity_scores_calib: NDArray,
+        **kwargs,
+    ) -> BaseCalibrator:
+        """
+        Fit the calibrator to estimate the conformity scores
+        quantiles. The method can take as arguments any of :
+        ``X, y, sample_weight, groups, y_pred_calib, conformity_scores_calib,
+        X_train, y_train, z_train, sample_weight_train, train_index,
+        X_calib, y_calib, z_calib, sample_weight_calib, calib_index``
+        or any other argument, which the user will have to pass as
+        ``**kwargs``.
+
+        Parameters
+        ----------
+        X_calib: ArrayLike of shape (n_samples, n_features)
+            Calibration data.
+
+        conformity_scores_calib: ArrayLike of shape (n_samples,)
+            Calibration conformity scores
+
+        Returns
+        -------
+        BaseCalibrator
+            Fitted self
+        """
+
```

Comment:
shouldn't we have `_calibrate` as abstract method for the base calibrator ?

Reply:
If we wanted to implement the ``CQR`` calibrator described in the [previous comment](https://github.com/scikit-learn-contrib/MAPIE/pull/455#discussion_r1648568238), as there is clearly a real ``fit`` part (fitting the ``alpha/2`` and ``1-alpha/2`` quantile estimator) and a ``calibrate`` part (conformalisation of the estimation), we could do something like:
```python
def fit(self, X_train, y_train, X_calib, y_calib, **kwargs):
      self._fit_estimator(self, X_train, y_train)
      self._calibrate(self, X_calib, y_calib)
```
But I guess most of the calibration methods only have 2 parts (calibration and prediction) so the base class probably don't need 3 by default.

Is the CQR example what you meant when you said to have a ``_calibrate`` method ? Or did I wrongly understood ?

Reply:
the idea was : you can give to MAPIE an unfitted model and in this case Mapie should handle the training. What is described is more a "calibrate" than a "fit" (even is a model can be fitted during the calibration)